### PR TITLE
[DATABEAM-133] Proxy for Spark Receivers

### PR DESCRIPTION
**Pull request by Amar3tto**
CDAP plugins that support streaming sources include custom Receiver classes (ex. `SalesforceReceiver)` that extend  abstract class `org.apache.spark.streaming.receiver.Receiver`. This class has few methods such as `store()`, `onStart()`, `onStop()` that use Spark environment.
As a step in integrating the Apache Beam and CDAP plugins, we need to have mechanisms to override this behavior and **not** use Spark inside Apache Beam. This can be done using the Proxy pattern, where we have a reference to the original custom `Receiver` object with the original behavior, but we can intercept all method calls and use our own implementation of some methods.
Untitled Diagram drawio(2)

For this purpose, suggested solution provides `ProxyBuilder<X, T extends Receiver<X>>` class. It is responsible for building Proxy objects of the Spark `Receiver` class.
Apache Beam user creates an instance of this class by passing the custom `Receiver` specific class as a `T` parameter. In this case, `X` parameter is a type of received data (e.x. `String`).
`ProxyBuilder<X, T extends Receiver<X>>` object provides simple methods to instantiate the corresponding `T` Proxy object:
1. `withConstructorArgs(Object... args)`
2. `withCustomStoreConsumer(Consumer<Object[]> storeConsumer)`
3. `withDefaultStoreConsumer()`
4. `build()`

Apache Beam user should specify required constructor parameters for instantiating custom Spark `Receiver` object. Then **_default_** or _**custom**_ realization of the `store()` method could be chosen. The main idea is that this realization will **not** use Spark environment and will use Apache Beam mechanisms instead.
After that, when the `build()` method was called, `ProxyBuilder` uses _CGLIB_ library to create and return ready-to-use Proxy object.
Since all Spark Receivers also use the Spark `ReceiverSupervisor` class, the behavior of this class must also be overridden. It was done using `WrappedSupervisor` class that extends `ReceiverSupervisor` and overrides all methods that could potentially use Spark Environment. Corresponding `WrappedSupervisor` object will be linked with the Proxy object that user built using `ProxyBuilder` class. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/io/cdap/src/test/java/org.apache.beam.sdk.io.cdap/ConfigWrapperTest.java [hunk 1]
```diff
@@ -53,27 +50,30 @@
...
     try {
       String newReferenceName = "new reference name";
-      SalesforceSourceConfig firstConfig = new ConfigWrapper<>(SalesforceSourceConfig.class)
-          .withParams(TEST_SALESFORCE_PARAMS_MAP)
-          .setParam("referenceName", newReferenceName)
-          .build();
+      SalesforceSourceConfig firstConfig =
+          new ConfigWrapper<>(SalesforceSourceConfig.class)
+              .withParams(TEST_SALESFORCE_PARAMS_MAP)
+              .setParam("referenceName", newReferenceName)
```
Comment (ktttnv) [inline sdks/java/io/cdap/src/test/java/org.apache.beam.sdk.io.cdap/ConfigWrapperTest.java:75]: I guess this variable may be used here.

```suggestion
              .setParam(REFERENCE_NAME_PARAM_NAME, newReferenceName)
```

File: sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+  public ProxyBuilder(Class<T> sparkReceiverClass) {
+    this.sparkReceiverClass = sparkReceiverClass;
+    withDefaultStoreConsumer();
+  }
+
+  /** Method for specifying constructor arguments for corresponding {@link #sparkReceiverClass} */
+  public ProxyBuilder<X, T> withConstructorArgs(Object... args) {
+    for (Constructor<?> constructor : sparkReceiverClass.getDeclaredConstructors()) {
+      if (constructor.getParameterCount() == args.length) {
```
[the remark this answers is not in the document]
Comment (Amar3tto) [reply sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java]: Used negative condition.

File: sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        boolean matches = true;
+        for (int i = 0; i < args.length; i++) {
+          Object arg = args[i];
+          if (arg == null) {
+            throw new IllegalArgumentException("All args must be not null!");
+          }
+          Class<?> currArgClass = constructor.getParameterTypes()[i];
```
[the remark this answers is not in the document]
Comment (Amar3tto) [reply sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java]: Done

File: sdks/java/io/cdap/src/test/java/org.apache.beam.sdk.io.cdap/ConfigWrapperTest.java (hunk 1 above)
Comment (Amar3tto) [reply sdks/java/io/cdap/src/test/java/org.apache.beam.sdk.io.cdap/ConfigWrapperTest.java]: Done

File: sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    MethodInterceptor handler =
+        (obj, method, args, proxy) -> {
+          String methodName = method.getName();
+          if (methodName.equals("supervisor") || methodName.equals("_supervisor")) {
+            return getWrappedSupervisor();
+          } else if (methodName.equals("onStart")) {
+            LOG.info("Custom Receiver was started");
+            return null;
+          } else if (methodName.equals("stop")) {
```
Comment (Lizzfox) [inline sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java:120]:
```suggestion
          } else if (methodName.equals("onStop")) {
```

Comment (Amar3tto) [reply sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java:120]: No, I meant `public void stop(String message, Throwable error)`

File: sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  /**
+   * Default realization of {@link Receiver#store(Object)} method, that uses Apache Beam mechanisms
+   * instead of Spark environment.
+   */
+  private void store(Object[] args) {
+    if (args.length == 1) {
+      Object arg = args[0];
+      if (arg instanceof Iterator) {
```
[the remark this answers is not in the document]
Comment (Amar3tto) [reply sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java:159]: Can't use switch for `instanceof`

File: sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    if (proxy != null) {
+      throw new IllegalStateException("Proxy already built.");
+    }
+    currentConstructor.setAccessible(true);
+    T originalReceiver = (T) currentConstructor.newInstance(constructorArgs);
+
+    MethodInterceptor handler =
+        (obj, method, args, proxy) -> {
+          String methodName = method.getName();
+          if (methodName.equals("supervisor") || methodName.equals("_supervisor")) {
```
[the remark this answers is not in the document]
Comment (Amar3tto) [reply sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/ProxyBuilder.java]: Done

**Review by Lizzfox — APPROVED** [review]

**Review by Krasavinigor — APPROVED** [review]

**Pull request closed without merging**